scripts/esim/esim.py

```diff
@@ -0,0 +1,124 @@
+        Dropout prob
+    """
+
+    def __init__(self, nwords, nword_dims, nhiddens, ndense_units,
+                 nclasses, drop_out, **kwargs):
+        super(ESIM, self).__init__(**kwargs)
+        with self.name_scope():
+            self.embedding_encoder = nn.Embedding(nwords, nword_dims)
+            self.batch_norm = nn.BatchNorm(axis=-1)
+            self.lstm_encoder1 = rnn.LSTM(nhiddens, bidirectional=True)
+            self.lstm_encoder2 = rnn.LSTM(nhiddens, bidirectional=True)
+
+            self.fc_encoder = nn.HybridSequential()
+            self.fc_encoder.add(nn.BatchNorm(axis=-1),
+                                nn.Dense(ndense_units),
+                                nn.ELU(),
+                                nn.BatchNorm(axis=-1),
+                                nn.Dropout(drop_out),
+                                nn.Dense(ndense_units),
+                                nn.ELU(),
+                                nn.BatchNorm(axis=-1),
+                                nn.Dropout(drop_out),
+                                nn.Dense(nclasses))
+
+            self.avg_pool = nn.GlobalAvgPool1D()
+            self.max_pool = nn.GlobalMaxPool1D()
+
+    def _soft_attention_align(self, F, x1, x2, mask1, mask2):
+        # attention shape: (batch, seq_len, seq_len)
+        attention = F.batch_dot(x1, x2, transpose_b=True)
+
+        weight1 = F.softmax(attention + F.expand_dims(mask2, axis=1), axis=-1)
+        x1_align = F.batch_dot(weight1, x2)
+        weight2 = F.softmax(attention + F.expand_dims(mask1, axis=1), axis=-1)
+        x2_align = F.batch_dot(weight2, x1)
+
+        return x1_align, x2_align
+
+    def _submul(self, F, x1, x2):
+        mul = F.multiply(x1, x2)
+        sub = F.subtract(x1, x2)
+
+        return F.concat(mul, sub, dim=-1)
+
+    def _apply_multiple(self, F, x):
+        p1 = F.squeeze(self.avg_pool(x), axis=-1)
+        p2 = F.squeeze(self.max_pool(x), axis=-1)
+
+        return F.concat(p1, p2, dim=-1)
+
+    def hybrid_forward(self, F, x1, x2, mask1, mask2):  # pylint: disable=arguments-differ
+        # x1_embed x2_embed shape: (batch, seq_len, nword_dims)
+        x1_embed = self.batch_norm(self.embedding_encoder(x1))
+        x2_embed = self.batch_norm(self.embedding_encoder(x2))
+
+        x1_lstm_encode = self.lstm_encoder1(x1_embed)
+        x2_lstm_encode = self.lstm_encoder1(x2_embed)
+
+        # attention
+        x1_algin, x2_algin = self._soft_attention_align(F, x1_lstm_encode, x2_lstm_encode,
+                                                        mask1, mask2)
+
+        # compose
+        x1_combined = F.concat(x1_lstm_encode, x1_algin,
+                               self._submul(F, x1_lstm_encode, x1_algin), dim=-1)
+        x2_combined = F.concat(x2_lstm_encode, x2_algin,
+                               self._submul(F, x2_lstm_encode, x2_algin), dim=-1)
+
+        x1_compose = self.lstm_encoder2(x1_combined)
+        x2_compose = self.lstm_encoder2(x2_combined)
```

Comment:
Should project `x1_combined` and `x2_combined` back to 300D before sending to LSTM.